### PR TITLE
update src folder files to ts

### LINKED ISSUE #4041: Convert to TS: src/checkValidTargetFilters.js
The following file `src/checkValidTargetFilters.js` should be converted to Typescript.

### LINKED ISSUE #4047: Convert to TS: src/accountImporter.js
The following file `src/accountImporter.js` should be converted to Typescript.

### LINKED ISSUE #4072: Convert to TS: src/index.js
The following file `src/index.js` should be converted to Typescript.

### LINKED ISSUE #4074: Convert to TS: src/responseToError.js
The following file `src/responseToError.js` should be converted to Typescript.

### LINKED ISSUE #4077: Convert to TS: src/parseBoltRules.js
The following file `src/parseBoltRules.js` should be converted to Typescript.

### LINKED ISSUE #4081: Convert to TS: src/functionsConfigClone.js
The following file `src/functionsConfigClone.js` should be converted to Typescript.

### PR DESCRIPTION
**Pull request by bkendall**
### Description

Rewrites most of the remaining files in `src/` to typescript.

Fixes #4081
Fixes #4077 
Fixes #4072 
Fixes #4074  
Fixes #4047 
Fixes #4041 

### Scenarios Tested

Deployed a few things, but making sure the integration tests pass will be the final test!

### CODE REVIEW

File: src/accountImporter.ts [hunk 1]
```diff
@@ -258,65 +250,65 @@ var _validateRequiredParameters = function (options) {
...
 
-var _validateProviderUserInfo = function (providerUserInfo) {
+function validateProviderUserInfo(providerUserInfo: any): any {
   if (!_.includes(ALLOWED_PROVIDER_IDS, providerUserInfo.providerId)) {
     return {
       error: JSON.stringify(providerUserInfo, null, 2) + " has unsupported providerId",
     };
   }
-  var keydiff = _.difference(_.keys(providerUserInfo), ALLOWED_PROVIDER_USER_INFO_KEYS);
+  const keydiff = _.difference(_.keys(providerUserInfo), ALLOWED_PROVIDER_USER_INFO_KEYS);
```
Comment (joehan) [inline src/accountImporter.ts:261]: Object.keys() over _.keys()

File: src/accountImporter.ts [hunk 2]
```diff
@@ -258,65 +250,65 @@ var _validateRequiredParameters = function (options) {
     default:
       throw new FirebaseError("Unsupported hash algorithm " + clc.bold(options.hashAlgo));
   }
-};
+}
 
-var _validateProviderUserInfo = function (providerUserInfo) {
+function validateProviderUserInfo(providerUserInfo: any): any {
```
Comment (joehan) [inline src/accountImporter.ts:255]: This can return `{ error?: string }` instead of any

File: src/accountImporter.ts [hunk 3]
```diff
@@ -258,65 +250,65 @@ var _validateRequiredParameters = function (options) {
...
         JSON.stringify(providerUserInfo, null, 2) + " has unsupported keys: " + keydiff.join(","),
     };
   }
   return {};
-};
+}
 
-var validateUserJson = function (userJson) {
-  var keydiff = _.difference(_.keys(userJson), ALLOWED_JSON_KEYS);
+export function validateUserJson(userJson: any): any {
```
Comment (joehan) [inline src/accountImporter.ts:271]: Same here - this can return `{ error?: string }`

File: src/accountImporter.ts [hunk 4]
```diff
@@ -328,21 +320,17 @@ var _sendRequest = function (projectId, userList, hashOptions) {
...
+}
 
-var serialImportUsers = function (projectId, hashOptions, userListArr, index) {
-  return _sendRequest(projectId, userListArr[index], hashOptions).then(function () {
+export function serialImportUsers(
+  projectId: string,
+  hashOptions: any,
+  userListArr: any[],
+  index: number
+): Promise<any> {
```
Comment (joehan) [inline src/accountImporter.ts:330]: I think this can return Promise<void>

File: src/checkValidTargetFilters.ts [hunk 5]
```diff
@@ -1,50 +1,43 @@
...
 
-var { FirebaseError } = require("./error");
-
-module.exports = function (options) {
-  function numFilters(targetTypes) {
-    return _.filter(options.only, function (opt) {
-      var optChunks = opt.split(":");
+export function checkValidTargetFilters(options: Options): Promise<void> {
+  function numFilters(targetTypes: string[]): number {
+    return _.filter(options.only, (opt) => {
```
Comment (joehan) [inline src/checkValidTargetFilters.ts:8]: Prefer native `options.only.filter`

File: src/functionsConfigClone.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    short,
+    (accum: boolean, x, i) => {
+      return accum && x === long[i];
+    },
+    true
+  );
+}
+
+function applyExcept(json: any, except: any[]) {
+  _.forEach(except, (key) => {
```
Comment (joehan) [inline src/functionsConfigClone.ts:23]: Prefer a normal for loop

File: src/functionsConfigClone.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  return runtimeconfig.variables.get(varName).then((variable) => {
+    const id = functionsConfig.varNameToIds(variable.name);
+    return runtimeconfig.variables.set(toProject, id.config, id.variable, variable.text);
+  });
+}
+
+function cloneConfig(configName: string, toProject: any): Promise<any> {
+  return runtimeconfig.variables.list(configName).then((variables) => {
+    return Promise.all(
+      _.map(variables, (variable) => {
```
Comment (joehan) [inline src/functionsConfigClone.ts:38]: Prefer native map

File: src/functionsConfigClone.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        return cloneVariable(variable.name, toProject);
+      })
+    );
+  });
+}
+
+async function cloneConfigOrVariable(key: string, fromProject: any, toProject: any): Promise<any> {
+  const parts = key.split(".");
+  if (_.includes(functionsConfig.RESERVED_NAMESPACES, parts[0])) {
```
Comment (joehan) [inline src/functionsConfigClone.ts:47]: Prefer native includes

File: src/functionsConfigClone.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    throw new FirebaseError("Cannot clone reserved namespace " + clc.bold(parts[0]));
+  }
+  const configName = _.join(["projects", fromProject, "configs", parts[0]], "/");
```
Comment (joehan) [inline src/functionsConfigClone.ts:50]: Prefer native join

File: src/functionsConfigClone.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  if (parts.length === 1) {
+    return cloneConfig(configName, toProject);
+  }
+  return runtimeconfig.variables.list(configName).then((variables) => {
+    const promises: Promise<any>[] = [];
+    _.forEach(variables, (variable) => {
```
Comment (joehan) [inline src/functionsConfigClone.ts:56]: Prefer normal for loop (or filter + map, in this case)

File: src/functionsConfigClone.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+export async function functionsConfigClone(
+  fromProject: any,
+  toProject: any,
+  only: string[] | undefined,
+  except: string[] = []
+): Promise<any> {
+  if (only) {
+    return Promise.all(
+      _.map(only, (key) => {
```
Comment (joehan) [inline src/functionsConfigClone.ts:75]: Native map over _.map

**Review by joehan — APPROVED** [review]
Again - bunch of other things that could be cleaned up here, but these are still progress toward cleaner code.

File: src/accountImporter.ts (hunk 1 above)
Comment (bkendall) [reply src/accountImporter.ts:261]: ack - will come back

File: src/accountImporter.ts (hunk 2 above)
Comment (bkendall) [reply src/accountImporter.ts:255]: Yes

File: src/accountImporter.ts (hunk 3 above)
Comment (bkendall) [reply src/accountImporter.ts:271]: wow, yeah, it can.

File: src/accountImporter.ts (hunk 4 above)
Comment (bkendall) [reply src/accountImporter.ts:330]: It is.

File: src/checkValidTargetFilters.ts (hunk 5 above)
Comment (bkendall) [reply src/checkValidTargetFilters.ts:8]: ack

File: src/functionsConfigClone.ts (hunk 6 above)
Comment (bkendall) [reply src/functionsConfigClone.ts:23]: ack

File: src/functionsConfigClone.ts (hunk 7 above)
Comment (bkendall) [reply src/functionsConfigClone.ts:38]: ack

File: src/functionsConfigClone.ts (hunk 8 above)
Comment (bkendall) [reply src/functionsConfigClone.ts:47]: ack

File: src/functionsConfigClone.ts (hunk 9 above)
Comment (bkendall) [reply src/functionsConfigClone.ts:50]: ack

File: src/functionsConfigClone.ts (hunk 10 above)
Comment (bkendall) [reply src/functionsConfigClone.ts:56]: ack

File: src/functionsConfigClone.ts (hunk 11 above)
Comment (bkendall) [reply src/functionsConfigClone.ts:75]: ack

**Pull request merged**